core/input/input.cpp

```diff
@@ -700,29 +711,39 @@ void Input::_parse_input_event_impl(const Ref<InputEvent> &p_event, bool p_is_em
 			continue;
 		}
 
-		Action &action = action_state[E.key];
+		int device_id = p_event->get_device();
+
+		ActionState &action_state = action_states[E.key];
+
 		if (!p_event->is_echo()) {
```

Comment:
The logic of echo events has changed. Previously even if event was echo it would still update strength, now it doesn't.
I didn't test if it has some essential consequences.

Reply:
Hmm, that's a good point. Probably I could use `!p_event->is_echo()` only for setting the "just pressed" state, as an echo event will simply update the corresponding event's `pressed` and `strength` states in the hashmap.
I can give it a try, just to avoid unexpected behaviors.

Reply:
I think the previous `is_echo()` check was done for performance reasons, as echo checks can't change pressed state, only strength. Now you removed it completely.

Though it should work correctly, so whatevs.